linkerd/proxy/detect/src/lib.rs

```diff
@@ -62,25 +47,63 @@ impl<D, A> Accept<D, A> {
     }
 }
 
-impl<T, D, A> tower::Service<(T, BoxedIo)> for Accept<D, A>
+impl<D: Clone, S> tower::layer::Layer<S> for DetectProtocolLayer<D> {
+    type Service = DetectProtocol<D, S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        DetectProtocol {
+            detect: self.detect.clone(),
+            peek_capacity: self.peek_capacity,
+            inner,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct DetectProtocol<D, S> {
+    detect: D,
+    peek_capacity: usize,
+    inner: S,
+}
+
+pub enum DetectProtocolFuture<T, D, S>
+where
+    S: tower::Service<(D::Target, BoxedIo)>,
+    D: Detect<T>,
+{
+    Detected(S::Future),
+    Detecting {
+        inner_svc: S,
+        detect: D,
+        inner: PeekAndDetect<T, D>,
+    },
+}
+
+impl<T, D, S> tower::Service<(T, BoxedIo)> for DetectProtocol<D, S>
 where
     D: Detect<T>,
-    A: core::listen::Accept<(D::Target, BoxedIo)> + Clone,
+    S: tower::Service<(D::Target, BoxedIo)>,
+    S: Clone,
+    S::Error: Into<Error>,
 {
-    type Response = ();
+    type Response = S::Response;
     type Error = Error;
-    type Future = AcceptFuture<T, D, A>;
+    type Future = DetectProtocolFuture<T, D, S>;
 
     fn poll_ready(&mut self) -> Poll<(), Self::Error> {
-        self.accept.poll_ready().map_err(Into::into)
+        self.inner.poll_ready().map_err(Into::into)
     }
 
     fn call(&mut self, (target, io): (T, BoxedIo)) -> Self::Future {
         match self.detect.detect_before_peek(target) {
-            Ok(detected) => AcceptFuture::Accept(self.accept.accept((detected, io))),
-            Err(target) => AcceptFuture::Detect {
+            Ok(detected) => {
+                let inner_fut = self.inner.call((detected, io));
+                return DetectProtocolFuture::Detected(inner_fut);
```

Comment:
nit: this isn't in a loop or assignment, so I don't think the `return` is necessary here
```suggestion
                DetectProtocolFuture::Detected(inner_fut)
```